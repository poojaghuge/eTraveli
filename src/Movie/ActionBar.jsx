import React, { useState } from "react";
import { styled, alpha } from "@mui/material/styles";
import {
  Paper,
  Box,
  Toolbar,
  InputBase,
  InputLabel,
  MenuItem,
  OutlinedInput,
} from "@mui/material";
import SearchIcon from "@mui/icons-material/Search";
import FormControl from "@mui/material/FormControl";
import Select from "@mui/material/Select";
import "./Movie.css";

const StyledSelect = styled(Select)(({ theme }) => ({
  color: "inherit",
  "& .MuiSelect-select, .MuiInputBase-input": {
    padding: theme.spacing(1, 1, 1, 0),
    paddingLeft: `${theme.spacing(4)})`,
    [theme.breakpoints.up("md")]: {
      width: "10ch",
    },
  },
}));



function ActionBar(props) {
  const [selectedValue, setSelectedValue] = useState("");

  const sortBy = ["Year", "Episode"];
  const handleSortChange = (event) => {
    const {
      target: { value },
    } = event;
    setSelectedValue(value);
    const order = value == "Year" ? "release_date" : "episode_id";
    props.handleSort(order);
  };

  return (
    <Paper sx={{ flexGrow: 1 }} variant={"outlined"}>
      <Toolbar>
        <FormControl sx={{ m: 1 }} className="sortBy">
          <StyledSelect
            labelId="sort-label"
            id="sort"
            displayEmpty
            value={selectedValue}
            onChange={handleSortChange}
            renderValue={(selectedValue) => {
              return <em>Sort by...{selectedValue}</em>;
            }}>
            <MenuItem disabled value="">
              <em>Sort by...</em>
            </MenuItem>
            {sortBy.map((name) => (
              <MenuItem key={name} value={name}>
                {name}
              </MenuItem>
            ))}
          </StyledSelect>
        </FormControl>
        <div className="search">
          <SearchIcon className="searchIcon" />
          <InputBase
            className="searchInput"
            placeholder="Type to search..."
            inputProps={{ "aria-label": "search" }}
            onChange={(event) => props.handleChange(event.target.value)}
          />
        </div>
        <Box sx={{ flexGrow: 1 }} />
      </Toolbar>
    </Paper>
  );
}

export const MemoisedActionBar = React.memo(ActionBar);
