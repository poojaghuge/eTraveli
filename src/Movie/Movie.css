.search {
  display: flex;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  border: 1px solid #f0f0f0e6;
  padding: 3px;
  border-radius: 5px;
}
.container{
    /* display: flex;
    justify-content: center;
    align-items: center; */
}
.searchIcon{
    color: #00000040;
}
.sortBy{
    color: #00000040
}
.searchInput{
    width:100%;
    padding-left: 1%;
}
.movieDetails{
    text-align: left;
}
.default{
    min-height: 60%;
    align-items: center;
    justify-content: center;
    display: flex;
    height: auto;
}