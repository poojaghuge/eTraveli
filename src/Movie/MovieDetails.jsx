import React, { useState, useEffect } from "react";
import { Box, Grid, Paper } from "@mui/material";

const MovieDetails = (props) => {
  const { selected } = props;
  if (!selected) {
    return (
      <div className="default">
        No Movie Selected
      </div>
    );
  }

  return (
    <div className="movieDetails">
      <h4>{selected?.title}</h4>
      <section>
        <div className="description" data-testid="description">
          {selected?.opening_crawl}
        </div>
        <br />
        <div data-testid="directedBy">Directed By: {selected?.director}</div>
      </section>
    </div>
  );
};

export const MemoisedMovieDetails = React.memo(MovieDetails);
