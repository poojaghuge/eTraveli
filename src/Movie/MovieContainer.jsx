import React, { useState, useEffect, useCallback } from "react";
import { Box, Grid, Paper } from "@mui/material";
import { getMovies } from "../common/service/MovieService";
import { MemoisedBasicTable } from "../common/components/BasicTable";
import { MemoisedMovieDetails } from "./MovieDetails";
import { MemoisedActionBar } from "./ActionBar";
import { sortArray } from "../common/utility/utilityMethods";

export const MovieContainer = () => {
  const [response, setResponse] = useState([]);
  const [data, setData] = useState([]);
  const [selected, setSelected] = useState();
  const [loading, setLoading] = useState(true);

  const getData = async () => {
    const response = await getMovies();
    setResponse(response.results);
    setData(response.results);
  };

  const showDetails = (item) => {
    setSelected(item);
  };

  const handleChange = (val) => {
    if (!val) {
      setData(response);
      return;
    }
    const result = data.filter((item) => {
      return item.title.toLowerCase().includes(val.toLowerCase());
    });
    setData(result);
  };

  const handleSort = (order) => {
    console.log(order);
    const array = sortArray([...data], order);
    console.log(array);
    setData(array);
  };

  useEffect(() => {
    getData();
    setLoading(false);
  }, []);

  const showDetailsMemo = useCallback(showDetails, []);
  const handleSortMemo = useCallback(handleSort, [data]);

  return (
    <>
      <MemoisedActionBar
        handleChange={handleChange}
        handleSort={handleSortMemo}
      />
      {loading ? (
        <div>Loading...</div>
      ) : (
        <Grid
          container
          rowSpacing={1}
          columnSpacing={{ xs: 1, sm: 2, md: 3 }}
          className="container">
          <Grid item xs={7}>
            <MemoisedBasicTable data={data} showDetails={showDetailsMemo} />
          </Grid>
          <Grid item xs={5}>
            <MemoisedMovieDetails selected={selected} />
          </Grid>
        </Grid>
      )}
    </>
  );
};
