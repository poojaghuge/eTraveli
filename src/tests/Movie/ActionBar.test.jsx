import { render, fireEvent } from "@testing-library/react";
import React from "react";
import renderer from "react-test-renderer";
import { MemoisedActionBar } from "../../Movie/ActionBar";

const handleChangeMock = jest.fn();
const handleSortMock = jest.fn();

describe("Test MemoisedActionBar", () => {
  it("test if matches the DOM Snapshot", () => {
    const domTree = renderer
      .create(
        <MemoisedActionBar
          handleChange={handleChangeMock}
          handleSort={handleSortMock}
        />
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });

  it("test if selected prop is passed then should render search & sort by", () => {
    const { getByPlaceholderText, getByText } = render(
      <MemoisedActionBar
        handleChange={handleChangeMock}
        handleSort={handleSortMock}
      />
    );
    const search = getByPlaceholderText("Type to search...");
    const sortBy = getByText("Sort by...");
    expect(search).toBeInTheDocument();
    expect(sortBy).toBeInTheDocument();
  });
});
