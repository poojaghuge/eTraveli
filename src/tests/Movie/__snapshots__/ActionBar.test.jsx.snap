// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MemoisedActionBar test if matches the DOM Snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded css-1sk68cw-MuiPaper-root"
>
  <div
    className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
  >
    <div
      className="MuiFormControl-root sortBy css-e363es-MuiFormControl-root"
    >
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-ddfktj-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        onClick={[Function]}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="sort-label sort"
          className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id="sort"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <em>
            Sort by...
          </em>
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          disabled={false}
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={false}
          tabIndex={-1}
          value=""
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-ihdtdm"
          >
            <span
              className="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="search"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium searchIcon css-i4bv87-MuiSvgIcon-root"
        data-testid="SearchIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <div
        className="MuiInputBase-root MuiInputBase-colorPrimary searchInput css-1le28he-MuiInputBase-root"
        onClick={[Function]}
      >
        <input
          aria-label="search"
          className="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Type to search..."
          type="text"
        />
      </div>
    </div>
    <div
      className="MuiBox-root css-i9gxme"
    />
  </div>
</div>
`;
